### PR TITLE
Rewrite screenshots.

### LINKED ISSUE #12478: Screenshots from any render target not just windows
## What problem does this solve or what need does it fill?

Example use cases:

1. If rendering to a non-window target we should be able to take screenshots of what's rendered.
2. In cases such as 4-player split screen we should be able to take individual screenshots of each split (likely covered by 1. if the split screen is implemented via first off-screen rendering then composited into a window)

## What solution would you like?

Allow taking screenshots of any render target: http://dev-docs.bevyengine.org/bevy/render/camera/enum.RenderTarget.html 

## What alternative(s) have you considered?

1. Spawning a dedicated window showing my offscreen rendered texture then taking a screenshot of that
2. Doing a manual texture-to-buffer copy then converting this to an image

## Additional context

_Just initial thoughts on how it might work_

From a quick look it seems that currently extracted windows have a special field: https://github.com/bevyengine/bevy/blob/main/crates/bevy_render/src/view/window/mod.rs#L73 

`pub screenshot_memory: Option<ScreenshotPreparedState>`

it seems to me that this shouldn't be necessary- perhaps a solution where something like

```rust
#[derive(..., Component)]
struct ScreenshotTarget {
  buf: Buffer
  ...
}
```

could be considered, and cameras (and windows?) with this component can have their screenshots taken.

This would also allow things like

```rust
// Screenshot with same width, height as camera target
commands.spawn((Camera3dBundle { ... }, ScreenshotTarget::default());

// Screenshot by sampling the camera no matter its size, then output to a FullHD sized image
commands.spawn((Camera3dBundle { ... }, ScreenshotTarget::with_size(1920, 1080));

// In both cases it doesn't matter if the cameras are 3D, 2D, if they render to separate windows, the same window, or not a window at all
```

There might be a loss of ergonomics if a user wants to take a screenshot of a window since they would have to add the `ScreenshotTarget` to the window entity (or entities), but an alternative could be to have windows spawn with this by default and rather have it opt-out.

### PR DESCRIPTION
**Pull request by tychedelia**
# Objective

Rewrite screenshotting to be able to accept any `RenderTarget`.

Closes #12478 

## Solution

Previously, screenshotting relied on setting a variety of state on the requested window. When extracted, the window's `swap_chain_texture_view` property would be swapped out with a texture_view created that frame for the screenshot pipeline to write back to the cpu.

Besides being tightly coupled to window in a way that prevented screenshotting other render targets, this approach had the drawback of relying on the implicit state of `swap_chain_texture_view` being returned from a `NormalizedRenderTarget` when view targets were prepared. Because property is set every frame for windows, that wasn't a problem, but poses a problem for render target images. Namely, to do the equivalent trick, we'd have to replace the `GpuImage`'s texture view, and somehow restore it later.

As such, this PR creates a new `prepare_view_textures` system which runs before `prepare_view_targets` that allows a new `prepare_screenshots` system to be sandwiched between and overwrite the render targets texture view if a screenshot has been requested that frame for the given target.

Additionally, screenshotting itself has been changed to use a component + observer pattern. We now spawn a `Screenshot` component into the world, whose lifetime is tracked with a series of marker components. When the screenshot is read back to the CPU, we send the image over a channel back to the main world where an observer fires on the screenshot entity before being despawned the next frame. This allows the user to access resources in their save callback that might be useful (e.g. uploading the screenshot over the network, etc.).

## Testing

image

TODO:
- [x] Web
- [ ] Manual texture view

---

## Showcase

render to texture example:

web saving still works:

## Migration Guide

`ScreenshotManager` has been removed. To take a screenshot, spawn a `Screenshot` entity with the specified render target and provide an observer targeting the `ScreenshotCaptured` event. See the `window/screenshot` example to see an example.

### CODE REVIEW

File: crates/bevy_render/src/view/window/screenshot.rs [hunk 1]
```diff
@@ -21,51 +18,80 @@ use crate::{
...
+use std::ops::Deref;
+use std::sync::mpsc::{Receiver, Sender};
+use std::sync::{Arc, Mutex};
+use std::{borrow::Cow, path::Path};
+use wgpu::{
+    CommandEncoder, Extent3d, ImageDataLayout, TextureFormat, COPY_BYTES_PER_ROW_ALIGNMENT,
 };
 
-use super::ExtractedWindows;
+#[derive(Event, Deref, DerefMut)]
```
Comment (killercup) [inline crates/bevy_render/src/view/window/screenshot.rs:40]: I see this wasn't the case before but maybe now is a good chance to derive Reflect on all of these? :)

File: crates/bevy_render/src/view/window/screenshot.rs [hunk 2]
```diff
@@ -221,114 +427,178 @@ impl SpecializedRenderPipeline for ScreenshotToScreenPipeline {
...
+            });
+            pass.set_pipeline(pipeline);
+            pass.set_bind_group(0, &prepared_state.bind_group, &[]);
+            pass.draw(0..3, 0..1);
         }
     }
 }
 
 pub(crate) fn collect_screenshots(world: &mut World) {
     let _span = info_span!("collect_screenshots");
```
Comment (killercup) [inline crates/bevy_render/src/view/window/screenshot.rs:552]: Not necessarily part of this PR but I think this span is never entered IIRC how tracing works. Maybe can be replaced with `#[intrument(level = "info", skip_all)]` on the fn?

File: examples/window/screenshot.rs [hunk 3]
```diff
@@ -13,17 +12,16 @@ fn main() {
...
 ) {
     if input.just_pressed(KeyCode::Space) {
         let path = format!("./screenshot-{}.png", *counter);
         *counter += 1;
-        screenshot_manager
-            .save_screenshot_to_disk(main_window.single(), path)
-            .unwrap();
+        commands
+            .spawn(Screenshot::primary_window())
+            .observe(save_to_disk(path));
```
Comment (killercup) [inline examples/window/screenshot.rs:24]: Would be really cool to extend this example to print out "screenshot saved to {}" on success to also demonstrate how to get that info.

**Review by killercup — COMMENTED** [review]
I've tried to follow the reasoning and the change and they make sense to me. I'm not confident in my skills to review the render changes.

**Comment by tychedelia** [comment]
I'm not totally sure if we should rewrite the headless renderer example to use this. It certainly would reduce a lot of boilerplate in that example but the example is also more efficient since it doesn't recreate textures and buffers for every screenshot like we do here. We could make this more efficient in engine by keeping those resources around rather than creating them for every new screenshot, but I'm not totally sure whether that's a use case we actually want to support in engine...

File: examples/window/screenshot.rs (hunk 3 above)
Comment (tychedelia) [reply examples/window/screenshot.rs:24]: We do print that out inside the `save_to_disk` callback... I think it's a helpful utility but anyone doing it "for real" probably just want's to roll their own. However, I think it's a good idea to demonstrate a different type of observer, so will do that 👍

File: crates/bevy_render/src/view/window/screenshot.rs (hunk 2 above)
Comment (tychedelia) [reply crates/bevy_render/src/view/window/screenshot.rs:583]: We don't seem to use the tracing proc macros in the engine for some reason, I'm not sure why. However, I'm pretty sure this should be guarded with `#[cfg(feature = "trace")]`, so thanks for pointing that out :p

File: crates/bevy_render/src/view/window/screenshot.rs (hunk 1 above)
Comment (tychedelia) [reply crates/bevy_render/src/view/window/screenshot.rs:40]: Okay, yeah, I'm never really sure when to derive reflect, will minimally put it on the event and screenshot component?

File: crates/bevy_render/src/view/window/screenshot.rs [hunk 4]
```diff
@@ -118,17 +180,215 @@ impl ScreenshotManager {
...
 impl Plugin for ScreenshotPlugin {
     fn build(&self, app: &mut bevy_app::App) {
-        app.init_resource::<ScreenshotManager>();
+        app.add_systems(
+            First,
+            clear_screenshots
+                .after(event_update_system)
+                .before(apply_deferred),
+        )
+        .add_systems(Update, trigger_screenshots);
```
Comment (killercup) [inline crates/bevy_render/src/view/window/screenshot.rs:391]: maybe i missed it but i think you need to register the reflected types here too

File: crates/bevy_render/src/view/window/screenshot.rs [hunk 5]
```diff
@@ -221,114 +496,179 @@ impl SpecializedRenderPipeline for ScreenshotToScreenPipeline {
...
+    let manual_texture_views = world.resource::<ManualTextureViews>();
+
+    for (entity, render_target) in targets.iter() {
+        match render_target {
+            NormalizedRenderTarget::Window(window) => {
+                let window = window.entity();
+                let window = windows.get(&window).unwrap();
+                let width = window.physical_width;
+                let height = window.physical_height;
+                let texture_format = window.swap_chain_texture_format.unwrap();
```
Comment (killercup) [inline crates/bevy_render/src/view/window/screenshot.rs:514]: Is this ever None? Or do we already know this is window always that has this?

File: examples/window/screenshot.rs [hunk 6]
```diff
@@ -1,29 +1,48 @@
...
+) {
+    let window = windows.single();
+    match screenshot_saving.iter().count() {
+        0 => {
+            commands.entity(window).remove::<CursorIcon>();
+        }
+        x if x > 0 => {
+            commands
+                .entity(window)
+                .insert(CursorIcon::from(SystemCursorIcon::Progress));
```
Comment (killercup) [inline examples/window/screenshot.rs:43]: Example works but -- as you know -- this cursor doesn't work on macos right now. Only mentioning here again to have a reference in https://github.com/rust-windowing/winit/issues/3724 that this is used :)

**Review by killercup — APPROVED** [review]
Example works, and code makes sense. Left a couple inline comments but consider this approved :)

File: crates/bevy_render/src/view/mod.rs [hunk 7]
```diff
@@ -809,24 +842,23 @@ pub fn prepare_view_targets(
...
-                    OutputColorAttachment::new(view.clone(), format.add_srgb_suffix())
-                })
-        }) else {
+        let Some(out_texture) = view_target_textures
+            .get(target)
+            .expect("Something went wrong with prepare_view_textures")
+            .as_ref()
+            .map(|(view, format)| {
+                OutputColorAttachment::new(view.clone(), format.add_srgb_suffix())
+            })
```
Comment (NthTensor) [inline crates/bevy_render/src/view/mod.rs:860]: I think there may be a subtle logic difference here. In the original, it looks if multiple cameras which render to the same target all use the same attachment. In the new version, every camera gets it's own attachment. At this point I'm not sure if that's going to effect how things work.

File: crates/bevy_render/src/view/mod.rs [hunk 8]
```diff
@@ -458,6 +465,11 @@ pub struct ViewTarget {
     out_texture: OutputColorAttachment,
 }
 
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct ViewTargetTextures(
```
Comment (NthTensor) [inline crates/bevy_render/src/view/mod.rs:469]: Please add documentation to this resource.

File: crates/bevy_render/src/view/mod.rs [hunk 9]
```diff
@@ -794,11 +806,32 @@ struct MainTargetTextures {
     main_texture: Arc<AtomicUsize>,
 }
 
-#[allow(clippy::too_many_arguments)]
-pub fn prepare_view_targets(
-    mut commands: Commands,
+pub fn prepare_view_textures(
```
Comment (NthTensor) [inline crates/bevy_render/src/view/mod.rs:809]: Nit: Very brief doc comments for both of thees systems would also be nice if you feel up to it.

**Review by NthTensor — COMMENTED** [review]
This PR is well motivated and I much prefer the new component/observer api. I am a little wary of the changes to `prepare_view_textures`, they seem generally like a big improvement but are not quite 1-to-1 with the behavior of the original.

In terms of reusing textures and porting over the headless example, I think that's something we can investigate as follow up. The perf and ergonomics are fine for now, and the screenshot example demonstrates use well enough.

File: crates/bevy_render/src/view/mod.rs (hunk 7 above)
Comment (tychedelia) [reply crates/bevy_render/src/view/mod.rs]: Cleaning this up actually simplified the code even more. Also added docs!

File: examples/window/screenshot.rs (hunk 6 above)
Comment (tychedelia) [reply examples/window/screenshot.rs:43]: Yeah, I'm leaving it in with the hope it eventually works 😭

File: crates/bevy_render/src/view/window/screenshot.rs (hunk 4 above)
Comment (tychedelia) [reply crates/bevy_render/src/view/window/screenshot.rs:400]: Ty, I'm a reflection noob.

File: crates/bevy_render/src/view/window/screenshot.rs (hunk 5 above)
Comment (tychedelia) [reply crates/bevy_render/src/view/window/screenshot.rs:523]: The scenario in which it's `None` is if you supply a `RenderTarget` with a window ref that's an invalid window entity. I made it so we don't extract those but I'll guard here just to be safe, unwraps are always a bit sketchy.

Comment (tychedelia) [reply crates/bevy_render/src/view/window/screenshot.rs]: Cleaned up a ton of unwraps which should be safe but u never know.

File: crates/bevy_render/src/view/mod.rs [hunk 10]
```diff
@@ -794,11 +808,41 @@ struct MainTargetTextures {
...
-#[allow(clippy::too_many_arguments)]
-pub fn prepare_view_targets(
-    mut commands: Commands,
+/// Prepares the view target [`OutputColorAttachment`] for each view in the current frame.
+pub fn prepare_view_attachments(
     windows: Res<ExtractedWindows>,
     images: Res<RenderAssets<GpuImage>>,
+    manual_texture_views: Res<ManualTextureViews>,
+    cameras: Query<&ExtractedCamera>,
+    mut view_target_textures: ResMut<ViewTargetAttachments>,
```
Comment (NthTensor) [inline crates/bevy_render/src/view/mod.rs:817]: name needs updating to match change to type.

File: crates/bevy_render/src/view/mod.rs [hunk 11]
```diff
@@ -809,24 +853,16 @@ pub fn prepare_view_targets(
...
 
-        let Some(out_texture) = output_textures.entry(target.clone()).or_insert_with(|| {
-            target
-                .get_texture_view(&windows, &images, &manual_texture_views)
-                .zip(target.get_texture_format(&windows, &images, &manual_texture_views))
-                .map(|(view, format)| {
-                    OutputColorAttachment::new(view.clone(), format.add_srgb_suffix())
-                })
-        }) else {
+        let Some(out_texture) = view_target_attachments.get(target) else {
```
Comment (NthTensor) [inline crates/bevy_render/src/view/mod.rs:865]: nit: technically this is an attachment not a texture, but it was incorrect before so it's not your fault. Still, mind fixing that?

**Review by NthTensor — APPROVED** [review]
Yep, looks like this preserves the old behavior now. Haven't run this code yet, but approving anyway.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1678 if you'd like to help out.

**Pull request merged**